File: test/cpu/cmpi.cc
#define BOOST_TEST_DYN_LINK
#include "68040.hpp"
#include "test.hpp"

#include <boost/test/data/monomorphic.hpp>
#include <boost/test/data/test_case.hpp>
#include <boost/test/unit_test.hpp>
namespace bdata = boost::unit_test::data;

BOOST_FIXTURE_TEST_SUITE(CMPI, Prepare)
struct F {
    F() {
        // CMPI.B #100, %D2 
        TEST::SET_W(0, 0006000 | 2);
        TEST::SET_W(2, 100);
        TEST::SET_W(4, TEST_BREAK);

        // CMPI.W #30000, %D2 
        TEST::SET_W(6, 0006100 | 2);
        TEST::SET_W(8, 30000);
        TEST::SET_W(10, TEST_BREAK);

        // CMPI.L #1000000000, %D2 
        TEST::SET_W(12, 0006200 | 2);
        TEST::SET_L(14, 1000000000);
        TEST::SET_W(18, TEST_BREAK);

        jit_compile(0, 20);
    }
};
BOOST_AUTO_TEST_SUITE(R, *boost::unit_test::fixture<F>())
BOOST_AUTO_TEST_SUITE(Byte)

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = 100;
    run_test(0);
    BOOST_TEST(cpu.Z);
}

BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = -10;
    run_test(0);
    BOOST_TEST(cpu.N);
}
BOOST_AUTO_TEST_CASE(C) {
    cpu.D[2] = 60;
    run_test(0);
    BOOST_TEST(cpu.C);
}
BOOST_AUTO_TEST_CASE(V) {
    cpu.D[2] = -120;
    run_test(0);
    BOOST_TEST(cpu.V);
}
BOOST_AUTO_TEST_SUITE_END()

BOOST_AUTO_TEST_SUITE(Word)

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = 30000;
    run_test(6);
    BOOST_TEST(cpu.Z);
}
BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = -100;
    run_test(6);
    BOOST_TEST(cpu.N);
}
BOOST_AUTO_TEST_CASE(C) {
    cpu.D[2] = 20000;
    run_test(6);
    BOOST_TEST(cpu.C);
}
BOOST_AUTO_TEST_CASE(V) {
    cpu.D[2] = -20000;
    run_test(6);
    BOOST_TEST(cpu.V);
}
BOOST_AUTO_TEST_SUITE_END()

BOOST_AUTO_TEST_SUITE(Long)


BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = 1000000000;
    run_test(12);
    BOOST_TEST(cpu.Z);
}
BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = -100;
    run_test(12);
    BOOST_TEST(cpu.N);
}
BOOST_AUTO_TEST_CASE(C) {
    cpu.D[2] = 900000000;
    run_test(12);
    BOOST_TEST(cpu.C);
}
BOOST_AUTO_TEST_CASE(V) {
    cpu.D[2] = -1500000000;
    run_test(12);
    BOOST_TEST(cpu.V);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()

File: test/lua/cpu/cmpi.lua
-- CMPI.B #10, %D1
MEM[0].W = OCT "006000" | 1
MEM[2].W = 10
MEM[4].W = TEST.BREAK

-- CMPI.W #30000, %D2
MEM[6].W = OCT "006100" | 2
MEM[8].H = 30000
MEM[10].W = TEST.BREAK

-- CMPI.L #1000000000, %D3
MEM[12].W = OCT "006200" | 3
MEM[14].I = 1000000000
MEM[18].W = TEST.BREAK

JIT_COMPILE(0, 20)
local run = {}
run.B = {
    z = farray(TF, function(b)
        CPU.D[2].C = cond(b, 10, 5)
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.Z", b)
    end),
    n = farray(TF, function(b)
        CPU.D[2].C = cond(b, -10, 100)
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.N", b)
    end),
    c = farray(TF, function(b)
        CPU.D[2].C = cond(b, 5, 20)
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.C", b)
    end),
    v = farray(TF, function(b)
        CPU.D[2].C = cond(b, -125, 20)
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.V", b)
    end)
}

run.W = {
    z = farray(TF, function(b)
        CPU.D[3].H = cond(b, 30000, 5)
        TEST.RUN(6)
        TEST.CHECK("CPU.CC.Z", b)
    end),
    n = farray(TF, function(b)
        CPU.D[3].H = cond(b, -100, 31000)
        TEST.RUN(6)
        TEST.CHECK("CPU.CC.N", b)
    end),
    c = farray(TF, function(b)
        CPU.D[3].H = cond(b, 20000, 31000)
        TEST.RUN(6)
        TEST.CHECK("CPU.CC.C", b)
    end),
    v = farray(TF, function(b)
        CPU.D[3].H = cond(b, -20000, 20)
        TEST.RUN(6)
        TEST.CHECK("CPU.CC.V", b)
    end)
}

run.L = {
    z = farray(TF, function(b)
        CPU.D[4].I = cond(b, 1000000000, 5)
        TEST.RUN(12)
        TEST.CHECK("CPU.CC.Z", b)
    end),
    n = farray(TF, function(b)
        CPU.D[4].I = cond(b, -100, 1200000000)
        TEST.RUN(12)
        TEST.CHECK("CPU.CC.N", b)
    end),
    c = farray(TF, function(b)
        CPU.D[4].I = cond(b, 900000000, 1200000000)
        TEST.RUN(12)
        TEST.CHECK("CPU.CC.C", b)
    end),
    v = farray(TF, function(b)
        CPU.D[4].I = cond(b, -1500000000, 20)
        TEST.RUN(12)
        TEST.CHECK("CPU.CC.V", b)
    end)
}

return run
